internal/controllers/clustertemplate_controller.go

```diff
@@ -239,6 +252,62 @@ func validateConfigmapReference[T any](
 	return nil
 }
 
+// validateName return true if the ClusterTemplate name is the
+// format: <name>.<version> or if the cluster <name> is used in
+// another namespace for a ClusterTemplate, false otherwise
+func validateName(c client.Client, name, version, metadataName, namespace string) error {
+	if metadataName != name+"."+version {
+		return utils.NewInputError("failed to validate ClusterTemplate name %s, should be in the format <spec.name>.<spec.version>: %s", metadataName, name+"."+version)
+	}
+
+	allClusterTemplates := &provisioningv1alpha1.ClusterTemplateList{}
+	err := c.List(context.Background(), allClusterTemplates)
+	if err != nil {
+		return fmt.Errorf("could not get list of ClusterTemplate across the cluster: %w", err)
+	}
+
+	sameMetadataName := map[string]bool{}
+	for _, aClusterTemplate := range allClusterTemplates.Items {
+		if aClusterTemplate.Namespace == namespace {
+			continue
+		}
+		if aClusterTemplate.Name == metadataName {
+			sameMetadataName[aClusterTemplate.Namespace] = true
+		}
+	}
+	if len(sameMetadataName) != 0 {
+		return utils.NewInputError("failed to validate ClusterTemplate name %s, a identical name already exists in namespaces: %s",
+			metadataName, strings.Join(utils.MapKeysToSlice(sameMetadataName), ","))
+	}
+	return nil
+}
+
+// validateTemplateID return true if the templateID is a valid uuid, false otherwise
+// If the templateID does not exist, it is generated by this function using uuid
+func validateTemplateID(ctx context.Context, c client.Client, object *provisioningv1alpha1.ClusterTemplate) error {
+	if object.Spec.TemplateID != "" {
+		_, err := uuid.Parse(object.Spec.TemplateID)
+		if err != nil {
+			return utils.NewInputError("failed to validate templateID, invalid UUID:%s", object.Spec.TemplateID)
+		}
+		return nil
+	}
+
+	newID := uuid.New()
+	newTemplate := object.DeepCopy()
+	newTemplate.Spec.TemplateID = newID.String()
+
+	err := utils.CreateK8sCR(ctx, c, newTemplate, object, utils.PATCH)
+	if err != nil {
+		return fmt.Errorf("failed to patch templateID in ClusterTemplate %s: %w", object.Name, err)
+	}
+	err = c.Get(ctx, types.NamespacedName{Name: object.Name, Namespace: object.Namespace}, object)
+	if err != nil {
+		return fmt.Errorf("failed to get updated ClusterTemplate %s: %w", object.Name, err)
+	}
```

Comment:
As we discussed, the status update issue caused by the race condition (where the resource hasn’t synced with the cache yet when querying immediately after an update) could be handled in a follow-up PR.